> OpenRA.Mods.Common/Traits/FieldDescriptionTooltip.cs(28,18): error CS0414: Warning as Error: The private field `OpenRA.Mods.Common.Traits.FieldDescriptionTooltip.self' is assigned but its value is never used

--- a/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj
+++ b/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj
@@ -303,6 +303,7 @@
     <Compile Include="Traits\CommandBarBlacklist.cs" />
     <Compile Include="Traits\Conditions\GrantRandomCondition.cs" />
     <Compile Include="Traits\Contrail.cs" />
+    <Compile Include="Traits\TooltipDescription.cs" />
     <Compile Include="Traits\Health.cs" />
     <Compile Include="Traits\HitShape.cs" />
     <Compile Include="Traits\Player\DeveloperMode.cs" />

--- a/OpenRA.Mods.Common/Traits/TooltipDescription.cs
+++ b/OpenRA.Mods.Common/Traits/TooltipDescription.cs
@@ -1,0 +1,41 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Additional info shown in the battlefield tooltip.")]
+	public class TooltipDescriptionInfo : ConditionalTraitInfo
+	{
+		[Desc("Text shown in tooltip.")]
+		[Translate]
+		public readonly string Description = "";
+
+		public override object Create(ActorInitializer init) { return new TooltipDescription(this); }
+	}
+
+	public class TooltipDescription : ConditionalTrait<TooltipDescriptionInfo>, IProvideTooltipInfo
+	{
+		public TooltipDescription(TooltipDescriptionInfo info)
+			: base(info) { }
+
+		public bool IsTooltipVisible(Player forPlayer) { return !IsTraitDisabled; }
+
+		public string TooltipText
+		{
+			get
+			{
+				return Info.Description;
+			}
+		}
+	}
+}

